### DISCUSSION TITLE
URL rewrites inside HTML content?

### DISCUSSION THREAD

**Discussion by marcosimioni** [gravitational/teleport#11536] [Q&A]
Hello,
I love teleport and it has a lot of cool features, but I'm struggling now with one specific application.

This application renders in its responses a bunch of `href`s having an hardcoded url, let's say `application.example.com:443`.
I'm using teleport to expose my application with the public URI `application.teleport.example.com:8443` (notice the domain and also the port number) and some areas work fine.
However, Teleport is not rewriting those `href`s values, so other features are not working because the browser of my end-user is trying to retrieve those assets via `application.example.com:443`, which of course does not work (remember? the application is exposed via `application.teleport.example.com:8443`).

The expected behavior would be that all references to `application.example.com:443` contained in the application responses should be rewritten to `application.teleport.example.com:8443`, or in other words what is described for instance here https://docs.microsoft.com/en-us/iis/extensions/url-rewrite-module/reverse-proxy-with-url-rewrite-v2-and-application-request-routing (`In addition, if internal application inserts links into its response HTML that link to elsewhere in those applications, those links should be modified before the response is returned to the client. For example, a page from http://webmail/default.aspx might contain a link like this:`).

Is there a way to enable such URL rewriting in Teleport? Or is teleport only able to rewrite URLs on HTTP forwards?

Thanks!

**Comment by marcosimioni** [comment]
I believe my issue is related to https://github.com/gravitational/teleport/issues/4716, but the proposed solution seems to work fine for HTTP 302 responses, but not for URLs located inside the HTML Content.

I'll investigate further, but if anybody has any pointer it would be more than welcome.

**Reply by webvictim** [reply]
I was the person who opened that issue as I had it too. I've just checked and it seems like my router software now uses whatever the URL in the browser is (i.e. `window.href` or similar) so it probably wasn't fixed in Teleport itself - it was just a coincidence.

I've reopened that issue and will try to investigate further. Teleport's reverse proxy support is unfortunately not as full-featured as that of other servers like nginx yet.

**Reply by marcosimioni** [reply]
Thanks a lot @webvictim for getting back to me. After looking at the source code, I realised the functionality is just missing.
I am working on a stub of functionality and it seems to be working ok.
I'll submit a PR soon, hopefully it will be merged at some stage.

**Reply by marcosimioni** [reply]
@webvictim I have submitted the PR (https://github.com/gravitational/teleport/pull/11769). Can you have a look at it by any chance, or is there someone else you know who could look into it?

I have no idea if I have forked from the right branch, nor if I have followed the process that Gravitational would like to be followed for contributions.

Thank you!

**Reply by webvictim** [reply]
Hey @marcosimioni, thanks for your work. Unfortunately it looks like we probably won't be accepting the PR upstream as-is, but hopefully we'll get to this when we look at plugin support for app access in future. For now, if the code changes work for you then I'm happy :)
